pkg/experiment/symbolizer/symbolizer.go

```diff
@@ -0,0 +1,448 @@
+			if !ok {
+				maxFuncID++
+				funcID = maxFuncID
+				profile.Function = append(profile.Function, &googlev1.Function{
+					Id:        funcID,
+					Name:      nameIdx,
+					Filename:  filenameIdx,
+					StartLine: int64(line.LineNumber),
+				})
+				funcMap[key] = funcID
+			}
+
+			profile.Location[locIdx].Line[j] = &googlev1.Line{
+				FunctionId: funcID,
+				Line:       int64(line.LineNumber),
+			}
+		}
+	}
+
+	mapping.HasFunctions = true
+	mapping.HasFilenames = true
+}
+
+func (s *Symbolizer) symbolize(ctx context.Context, req *request) error {
+	start := time.Now()
+	status := statusSuccess
+	defer func() {
+		s.metrics.profileSymbolization.WithLabelValues(status).Observe(time.Since(start).Seconds())
+	}()
+
+	var table *lidia.Table
+	var err error
+
+	lidiaBytes, err := s.getLidiaBytes(ctx, req.buildID)
+	if err != nil {
+		var bnfErr buildIDNotFoundError
+		if errors.As(err, &bnfErr) {
+			for _, loc := range req.locations {
+				loc.lines = s.createNotFoundSymbols(req.binaryName, loc)
+			}
+			return nil
+		}
+
+		status = statusErrorServerError
+		return err
+	}
+
+	lidiaReader := NewReaderAtCloser(lidiaBytes)
+	table, err = lidia.OpenReader(lidiaReader, lidia.WithCRC())
+	if err != nil {
+		s.metrics.debugSymbolResolution.WithLabelValues("lidia_error").Observe(0)
+		status = statusErrorServerError
+		return fmt.Errorf("open lidia file: %w", err)
+	}
+	defer table.Close()
+
+	s.symbolizeWithTable(ctx, table, req)
+	return nil
+}
+
+func (s *Symbolizer) symbolizeWithTable(_ context.Context, table *lidia.Table, req *request) {
+	var framesBuf []lidia.SourceInfoFrame
+
+	resolveStart := time.Now()
+	defer func() {
+		s.metrics.debugSymbolResolution.WithLabelValues(statusSuccess).Observe(time.Since(resolveStart).Seconds())
+	}()
+
+	for _, loc := range req.locations {
+		frames, err := table.Lookup(framesBuf, loc.address)
+		if err != nil {
+			loc.lines = s.createNotFoundSymbols(req.binaryName, loc)
+			continue
+		}
+
+		if len(frames) == 0 {
+			loc.lines = s.createNotFoundSymbols(req.binaryName, loc)
+			continue
+		}
+
+		loc.lines = frames
+	}
+}
+
+func (s *Symbolizer) getLidiaBytes(ctx context.Context, buildID string) ([]byte, error) {
+	lidiaBytes, err := s.fetchLidiaFromObjectStore(ctx, buildID)
+	if err == nil {
+		return lidiaBytes, nil
+	}
+
+	level.Error(s.logger).Log(
+		"msg", "failed to get from object store",
+		"build_id", buildID,
+		"error", err,
+	)
+
+	lidiaBytes, err = s.fetchLidiaFromDebuginfod(ctx, buildID)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := s.store.Put(ctx, buildID, bytes.NewReader(lidiaBytes)); err != nil {
+		level.Warn(s.logger).Log("msg", "Failed to store debug info in objstore", "buildID", buildID, "err", err)
+	}
+
+	return lidiaBytes, nil
+}
+
+// fetchLidiaFromObjectStore retrieves Lidia data from the object store
+func (s *Symbolizer) fetchLidiaFromObjectStore(ctx context.Context, buildID string) ([]byte, error) {
+	objstoreReader, err := s.store.Get(ctx, buildID)
```

Comment:
can we avoid store.Get invocations if we already know debuginfod returned NotFound ? it would probably require to pull the cache from debuginfod into symbolizer